pkg/dss/scd/cockroach/subscriptions.go

```diff
@@ -0,0 +1,445 @@
+
+	"github.com/golang/geo/s2"
+	"github.com/lib/pq"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+const (
+	// Defined in requirement DSS0030.
+	maxSubscriptionsPerArea = 10
+)
+
+var (
+	subscriptionFieldsWithIndices   [11]string
+	subscriptionFieldsWithPrefix    string
+	subscriptionFieldsWithoutPrefix string
+)
+
+func init() {
+	subscriptionFieldsWithIndices[0] = "id"
+	subscriptionFieldsWithIndices[1] = "owner"
+	subscriptionFieldsWithIndices[2] = "version"
+	subscriptionFieldsWithIndices[3] = "url"
+	subscriptionFieldsWithIndices[4] = "notification_index"
+	subscriptionFieldsWithIndices[5] = "notify_for_operations"
+	subscriptionFieldsWithIndices[6] = "notify_for_constraints"
+	subscriptionFieldsWithIndices[7] = "implicit"
+	subscriptionFieldsWithIndices[8] = "starts_at"
+	subscriptionFieldsWithIndices[9] = "ends_at"
+	subscriptionFieldsWithIndices[10] = "updated_at"
+
+	subscriptionFieldsWithoutPrefix = strings.Join(
+		subscriptionFieldsWithIndices[:], ",",
+	)
+
+	withPrefix := make([]string, 11)
+	for idx, field := range subscriptionFieldsWithIndices {
+		withPrefix[idx] = "scd_subscriptions." + field
+	}
+
+	subscriptionFieldsWithPrefix = strings.Join(
+		withPrefix[:], ",",
+	)
+}
+
+func (c *Store) fetchSubscriptions(ctx context.Context, q queryable, query string, args ...interface{}) ([]*scdmodels.Subscription, error) {
+	rows, err := q.QueryContext(ctx, query, args...)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var payload []*scdmodels.Subscription
+	for rows.Next() {
+		var (
+			s         = new(scdmodels.Subscription)
+			updatedAt time.Time
+		)
+		err := rows.Scan(
+			&s.ID,
+			&s.Owner,
+			&s.Version,
+			&s.BaseURL,
+			&s.NotificationIndex,
+			&s.NotifyForOperations,
+			&s.NotifyForConstraints,
+			&s.ImplicitSubscription,
+			&s.StartTime,
+			&s.EndTime,
+			&updatedAt,
+		)
+		if err != nil {
+			return nil, err
+		}
+		payload = append(payload, s)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+	return payload, nil
+}
+
+func (c *Store) fetchSubscriptionsForNotification(
+	ctx context.Context, q queryable, cells []int64) ([]*scdmodels.Subscription, error) {
+	// TODO(dsansome): upgrade to cockroachdb 19.2.0 and convert this to a single
+	// UPDATE FROM query.
+
+	// First: get unique subscription IDs.
+	var query = `
+			SELECT DISTINCT
+				subscription_id
+			FROM
+				scd_cells_subscriptions
+			WHERE
+				cell_id = ANY($1)`
+	rows, err := q.QueryContext(ctx, query, pq.Array(cells))
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var subscriptionIDs []string
+	for rows.Next() {
+		var id string
+		if err := rows.Scan(&id); err != nil {
+			return nil, err
+		}
+		subscriptionIDs = append(subscriptionIDs, id)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+
+	// Next: update the notification_index of each one and return the rest of the
+	// data.
+	var updateQuery = fmt.Sprintf(`
+			UPDATE
+				scd_subscriptions
```

Comment:
FYI: Not for this review, but I found a library called squirrel which is nice https://github.com/Masterminds/squirrel

Reply:
Added #262